File: analyze_index.py
#!/usr/bin/env python3
import json
import re
import base64

# Read the root.js file
with open('target/doc/search.index/root.js', 'r', encoding='utf-8') as f:
    content = f.read()

# Extract JSON from rr_('...')
match = re.search(r"rr_\('(.+)'\)", content, re.DOTALL)
if not match:
    print("Could not find rr_('...') pattern")
    exit(1)

json_str = match.group(1)

# Parse JSON
data = json.loads(json_str)

print("=== ROOT STRUCTURE ===\n")
for key, value in data.items():
    print(f"Key: {key}")
    if isinstance(value, dict):
        print(f"  Type: Object")
        for k, v in value.items():
            v_type = type(v).__name__
            if isinstance(v, str):
                print(f"    - {k}: {v_type} (length: {len(v)})")
            else:
                print(f"    - {k}: {v_type}")
    elif isinstance(value, list):
        print(f"  Type: List (length: {len(value)})")
    elif isinstance(value, str):
        print(f"  Type: String (length: {len(value)})")
    print()

# Analyze the compressed fields
print("\n=== ANALYZING COMPRESSED FIELDS ===\n")

for field_name in ['normalizedName', 'name', 'path', 'entry', 'desc', 'function', 'type']:
    if field_name in data:
        print(f"{field_name} structure:")
        field = data[field_name]
        if isinstance(field, dict):
            for key, value in field.items():
                if isinstance(value, str):
                    print(f"  {key}: string (length: {len(value)})")
                    if len(value) < 100:
                        print(f"      Value: {value}")
                else:
                    print(f"  {key}: {type(value).__name__}")
        print()

# Try to understand the encoding
print("\n=== DECODING ATTEMPT ===\n")

if 'normalizedName' in data and 'I' in data['normalizedName']:
    i_field = data['normalizedName']['I']
    print(f"normalizedName.I (first 200 chars):")
    print(i_field[:200])
    print()

    # Try to decode as base64
    try:
        decoded = base64.b64decode(i_field)
        print(f"Base64 decode successful! Length: {len(decoded)} bytes")
        print(f"First 100 bytes as hex:")
        print(' '.join(f'{b:02x}' for b in decoded[:100]))
        print()
        print(f"First 100 bytes as ASCII (. for non-printable):")
        print(''.join(chr(b) if 32 <= b < 127 else '.' for b in decoded[:100]))
        print()

        # Try to understand the structure
        # Look for patterns - maybe it's variable-length encoded
        print("Looking for null-terminated strings...")
        strings = []
        current = []
        for b in decoded[:500]:
            if b == 0:
                if current:
                    s = bytes(current).decode('utf-8', errors='ignore')
                    if s and len(s) > 1:
                        strings.append(s)
                    current = []
            elif 32 <= b < 127:  # printable ASCII
                current.append(b)
            else:
                if current:
                    s = bytes(current).decode('utf-8', errors='ignore')
                    if s and len(s) > 1:
                        strings.append(s)
                    current = []

        if strings:
            print(f"Found {len(strings)} potential strings:")
            for s in strings[:20]:  # First 20
                print(f"  - {s}")
    except Exception as e:
        print(f"Base64 decode failed: {e}")

# Check what N, E, H fields are
print("\n\n=== N, E, H FIELDS ===\n")
if 'normalizedName' in data:
    field = data['normalizedName']
    if 'N' in field:
        print(f"N field: {field['N']}")
    if 'E' in field:
        print(f"E field (length {len(field['E'])}): {field['E'][:100] if len(field['E']) > 100 else field['E']}")
        try:
            e_decoded = base64.b64decode(field['E'])
            print(f"  E decoded: {len(e_decoded)} bytes")
            print(f"  As integers: {list(e_decoded[:20])}")
        except:
            pass
    if 'H' in field:
        print(f"H field: {field['H']}")
        print(f"  Looks like a hash: {len(field['H'])} chars")

print("\n\n=== SAMPLE DATA FROM DIFFERENT FIELDS ===\n")

# Print samples from each compressed field to see patterns
for field_name in ['normalizedName', 'name', 'path']:
    if field_name in data:
        field = data[field_name]
        if 'I' in field and isinstance(field['I'], str):
            print(f"{field_name}.I first 300 chars:")
            print(field['I'][:300])
            print()

File: analyze_index.rs
use serde_json::Value;
use std::fs;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Read the root.js file
    let content = fs::read_to_string("target/doc/search.index/root.js")?;

    // Extract JSON from rr_('...')
    let start = content.find("rr_('").unwrap() + 5;
    let end = content.rfind("')").unwrap();
    let json_str = &content[start..end];

    // Parse JSON
    let data: Value = serde_json::from_str(json_str)?;

    println!("=== ROOT STRUCTURE ===\n");
    if let Value::Object(map) = &data {
        for (key, value) in map {
            println!("Key: {}", key);
            match value {
                Value::Object(obj) => {
                    println!("  Type: Object");
                    for (k, v) in obj {
                        println!("    - {}: {}", k, describe_value(v));
                    }
                }
                Value::Array(arr) => {
                    println!("  Type: Array (length: {})", arr.len());
                    if !arr.is_empty() {
                        println!("    First element: {}", describe_value(&arr[0]));
                    }
                }
                Value::String(s) => {
                    println!("  Type: String (length: {})", s.len());
                    if s.len() < 100 {
                        println!("    Value: {}", s);
                    } else {
                        println!("    Value: {}...", &s[..100]);
                    }
                }
                _ => println!("  Type: {:?}", value),
            }
            println!();
        }
    }

    // Analyze the compressed fields
    println!("\n=== ANALYZING COMPRESSED FIELDS ===\n");

    if let Some(obj) = data.as_object() {
        // Look at normalizedName structure
        if let Some(normalized_name) = obj.get("normalizedName") {
            println!("normalizedName structure:");
            analyze_compressed_field(normalized_name);
        }

        if let Some(name) = obj.get("name") {
            println!("\nname structure:");
            analyze_compressed_field(name);
        }

        if let Some(path) = obj.get("path") {
            println!("\npath structure:");
            analyze_compressed_field(path);
        }

        if let Some(entry) = obj.get("entry") {
            println!("\nentry structure:");
            analyze_compressed_field(entry);
        }

        if let Some(desc) = obj.get("desc") {
            println!("\ndesc structure:");
            analyze_compressed_field(desc);
        }
    }

    // Try to decode the base64-like strings
    println!("\n=== ATTEMPTING TO DECODE ===\n");

    if let Some(obj) = data.as_object() {
        if let Some(normalized_name) = obj.get("normalizedName") {
            if let Some(i_field) = normalized_name.get("I") {
                if let Some(i_str) = i_field.as_str() {
                    println!("normalizedName.I (first 200 chars):");
                    println!("{}", &i_str[..i_str.len().min(200)]);
                    println!("\nAttempting to decode as base64...");

                    // Try decoding
                    if let Ok(decoded) = base64_decode_custom(i_str) {
                        println!("Decoded length: {} bytes", decoded.len());
                        println!("First 50 bytes as hex: {}",
                            decoded.iter()
                                .take(50)
                                .map(|b| format!("{:02x}", b))
                                .collect::<Vec<_>>()
                                .join(" "));
                        println!("First 50 bytes as ASCII (. for non-printable): {}",
                            decoded.iter()
                                .take(50)
                                .map(|&b| if b >= 32 && b < 127 { b as char } else { '.' })
                                .collect::<String>());
                    }
                }
            }
        }
    }

    Ok(())
}

fn describe_value(value: &Value) -> String {
    match value {
        Value::Null => "null".to_string(),
        Value::Bool(b) => format!("bool: {}", b),
        Value::Number(n) => format!("number: {}", n),
        Value::String(s) => {
            if s.len() < 50 {
                format!("string: \"{}\"", s)
            } else {
                format!("string (len {}): \"{}...\"", s.len(), &s[..47])
            }
        }
        Value::Array(a) => format!("array (len {})", a.len()),
        Value::Object(o) => format!("object (keys: {})", o.keys().map(|k| k.as_str()).collect::<Vec<_>>().join(", ")),
    }
}

fn analyze_compressed_field(field: &Value) {
    if let Value::Object(obj) = field {
        for (key, value) in obj {
            println!("  {}: {}", key, describe_value(value));
        }
    }
}

fn base64_decode_custom(s: &str) -> Result<Vec<u8>, String> {
    // This appears to be a custom base64-like encoding
    // Standard base64 alphabet: ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/

    use base64::{Engine as _, engine::general_purpose};
    general_purpose::STANDARD.decode(s).map_err(|e| e.to_string())
}
